### ISSUE TITLE
persistence:  100% CPU loop after restarting with 50 tables 

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#7969]
### What version of Materialize are you using?

```
v0.9.1-dev (780a805fd) 
```

### What was the issue?

Restarting with 50 persisted tables never completes.

It takes progressively longer as the number of tables is increased starting from 1, but it seems to be more than a high-O-complexity algorithm going on, I think it is looping permanently after some threshold.

perf reports that materialize is stuck like this:

```
-  100.00%     0.00%  materialized  [unknown]          [.] 0x0000000000000001                                                                                                                                ▒
   - 0x1                                                                                                                                                                                                     ▒
        58.22% <futures_executor::local_pool::ThreadNotify as futures_task::arc_wake::ArcWake>::wake_by_ref                                                                                                  ▒
        23.44% std::thread::local::LocalKey<T>::with                                                                                                                                                         ▒
        8.48% <tokio::sync::oneshot::Receiver<T> as core::future::future::Future>::poll                                                                                                                      ◆
        4.17% <futures_util::future::future::map::Map<Fut,F> as core::future::future::Future>::poll                                                                                                          ▒
        2.95% tokio::coop::Budget::decrement                                                                                                                                                                 ▒
        2.74% futures_task::waker::wake_by_ref_arc_raw                                                                                                                                                       ▒
+   58.22%    58.12%  materialized  materialized       [.] <futures_executor::local_pool::ThreadNotify as futures_task::arc_wake::ArcWake>::wake_by_ref                                                      ▒
+   23.44%    23.42%  materialized  materialized       [.] std::thread::local::LocalKey<T>::with                                                                                                             ▒
+    8.48%     8.47%  materialized  materialized       [.] <tokio::sync::oneshot::Receiver<T> as core::future::future::Future>::poll                                                                         ▒
+    4.17%     4.16%  materialized  materialized       [.] <futures_util::future::future::map::Map<Fut,F> as core::future::future::Future>::poll                                                             ▒
+    2.95%     2.95%  materialized  materialized       [.] tokio::coop::Budget::decrement                                                                                                                    ▒
+    2.74%     2.74%  materialized  materialized       [.] futures_task::waker::wake_by_ref_arc_raw                                                            

```

Initially, strace reports some I/O against `mzdata/persist/blob/META*`  but then no further syscalls are issued.

### Is the issue reproducible? If so, please provide reproduction instructions.

Pipe the output of this script to the database:

```
for i in range (0,50):
   print(f"CREATE TABLE t{i} (f1 INTEGER);")
   print(f"INSERT INTO t{i} VALUES (1);")
   print()
```

**Comment by philip-stoev** [comment]
It always seems to hang in the same place:

```
...
2021-08-20T08:28:45.997058Z  INFO coord::catalog: create index materialize.public.t48_primary_idx (u98)    
2021-08-20T08:28:46.016556Z  INFO coord::catalog: create table materialize.public.t49 (u99)    
2021-08-20T08:28:46.036922Z  INFO coord::catalog: create index materialize.public.t49_primary_idx (u100)    
2021-08-20T08:28:46.065539Z  INFO coord::catalog: create table materialize.public.t0 (u1)    
2021-08-20T08:28:46.085274Z  INFO coord::catalog: create index materialize.public.t0_primary_idx (u2)    
2021-08-20T08:28:46.104687Z  INFO coord::catalog: create table materialize.public.t1 (u3)    
2021-08-20T08:28:46.124511Z  INFO coord::catalog: create index materialize.public.t1_primary_idx (u4)    
2021-08-20T08:28:46.143730Z  INFO coord::catalog: create table materialize.public.t2 (u5)    
2021-08-20T08:28:46.163298Z  INFO coord::catalog: create index materialize.public.t2_primary_idx (u6)    
2021-08-20T08:28:46.183353Z  INFO coord::catalog: create table materialize.public.t3 (u7)    
2021-08-20T08:28:46.205420Z  INFO coord::catalog: create index materialize.public.t3_primary_idx (u8)    
2021-08-20T08:28:46.225552Z  INFO coord::catalog: create table materialize.public.t4 (u9)    
2021-08-20T08:28:46.245344Z  INFO coord::catalog: create index materialize.public.t4_primary_idx (u10)    
2021-08-20T08:28:46.265065Z  INFO coord::catalog: create table materialize.public.t5 (u11)    
2021-08-20T08:28:46.284301Z  INFO coord::catalog: create index materialize.public.t5_primary_idx (u12)    
2021-08-20T08:28:46.304284Z  INFO coord::catalog: create table materialize.public.t6 (u13)    
2021-08-20T08:28:46.326996Z  INFO coord::catalog: create index materialize.public.t6_primary_idx (u14)    
2021-08-20T08:28:46.346642Z  INFO coord::catalog: create table materialize.public.t7 (u15)    
2021-08-20T08:28:46.366830Z  INFO coord::catalog: create index materialize.public.t7_primary_idx (u16)    
2021-08-20T08:28:46.386548Z  INFO coord::catalog: create table materialize.public.t8 (u17)    
2021-08-20T08:28:46.407082Z  INFO coord::catalog: create index materialize.public.t8_primary_idx (u18)    
2021-08-20T08:28:46.426563Z  INFO coord::catalog: create table materialize.public.t9 (u19)    
2021-08-20T08:28:46.446606Z  INFO coord::catalog: create index materialize.public.t9_primary_idx (u20)    
2021-08-20T08:28:46.466273Z  INFO coord::catalog: create table materialize.public.t10 (u21)    
2021-08-20T08:28:46.486555Z  INFO coord::catalog: create index materialize.public.t10_primary_idx (u22)    
2021-08-20T08:28:46.506515Z  INFO coord::catalog: create table materialize.public.t11 (u23)    
2021-08-20T08:28:46.526266Z  INFO coord::catalog: create index materialize.public.t11_primary_idx (u24)    
2021-08-20T08:28:46.546573Z  INFO coord::catalog: create table materialize.public.t12 (u25)    
2021-08-20T08:28:46.565731Z  INFO coord::catalog: create index materialize.public.t12_primary_idx (u26)    
2021-08-20T08:28:46.597778Z  INFO coord::catalog: create table materialize.public.t13 (u27)    
2021-08-20T08:28:46.617508Z  INFO coord::catalog: create index materialize.public.t13_primary_idx (u28)    
...
```

So when the 14-th table is mentioned on the second round of "create" messages.

**Comment by ruchirK** [comment]
this is really odd. looking now

**Comment by ruchirK** [comment]
jotting down notes here:

this seems like a bug where we create a future / future handle send it to runtime, and then the runtime processes the future correctly, and and sends a response back, but the runtime client never receives the response

I added some hopefully self describing logging statements:

```
2021-08-25T00:32:09.938707Z  INFO coord::catalog: create index materialize.public.t11_primary_idx (u24)
2021-08-25T00:32:09.938687Z  WARN persist::indexed: drain pending called
2021-08-25T00:32:09.938837Z  WARN persist::indexed::runtime: sending create for id "user-table-25-materialize.public.t12"
2021-08-25T00:32:09.938849Z  WARN persist::indexed::runtime: awaiting receipt of create for id "user-table-25-materialize.public.t12"
2021-08-25T00:32:10.041554Z  WARN persist::indexed::runtime: top of work
2021-08-25T00:32:10.041582Z  WARN persist::indexed: register called for id "user-table-25-materialize.public.t12"
2021-08-25T00:32:10.041592Z  WARN persist::indexed: drain pending called
2021-08-25T00:32:10.121714Z  WARN persist::indexed::runtime: stepping
2021-08-25T00:32:10.121764Z  WARN persist::indexed: drain pending called
2021-08-25T00:32:10.121771Z  WARN persist::indexed::runtime: received ack of create for id Id(14)
2021-08-25T00:32:10.121797Z  INFO coord::catalog: create table materialize.public.t12 (u25)
2021-08-25T00:32:10.121948Z  WARN persist::indexed::runtime: sending create for id "user-table-26-materialize.public.t12_primary_idx"
2021-08-25T00:32:10.121965Z  WARN persist::indexed::runtime: awaiting receipt of create for id "user-table-26-materialize.public.t12_primary_idx"
2021-08-25T00:32:10.192645Z  WARN persist::indexed::runtime: top of work
2021-08-25T00:32:10.192672Z  WARN persist::indexed: register called for id "user-table-26-materialize.public.t12_primary_idx"
2021-08-25T00:32:10.192682Z  WARN persist::indexed: drain pending called
2021-08-25T00:32:10.240875Z  WARN persist::indexed::runtime: stepping
2021-08-25T00:32:10.240909Z  WARN persist::indexed: drain pending called
2021-08-25T00:32:10.240934Z  WARN persist::indexed::runtime: received ack of create for id Id(64)
2021-08-25T00:32:10.240969Z  INFO coord::catalog: create index materialize.public.t12_primary_idx (u26)
2021-08-25T00:32:10.241139Z  WARN persist::indexed::runtime: sending create for id "user-table-27-materialize.public.t13"
2021-08-25T00:32:10.241151Z  WARN persist::indexed::runtime: awaiting receipt of create for id "user-table-27-materialize.public.t13"
2021-08-25T00:32:10.309879Z  WARN persist::indexed::runtime: top of work
2021-08-25T00:32:10.309910Z  WARN persist::indexed: register called for id "user-table-27-materialize.public.t13"
2021-08-25T00:32:10.309915Z  WARN persist::indexed: drain pending called
2021-08-25T00:32:10.359524Z  WARN persist::indexed::runtime: stepping
2021-08-25T00:32:10.359536Z  WARN persist::indexed: drain pending called
2021-08-25T00:32:10.498529Z  WARN persist::indexed::runtime: top of work
```

This is also corroborated by profiling, the persist runtime thread is spending all of its time waiting for the next message, and the coordinator thread is blocked on awaiting the response of the future.

Also, note that we create 2x the number of persistent streams here -- one for the tables, and one for the indexes. That's unintentional and might be related to the hang or it might be incidental

**Comment by ruchirK** [comment]
Ok. I just pushed a commit that "fixes" this issue in that, we don't hang indefinitely when we attempt to restart. Two things to note:

- its not clear if I've actually addressed the root cause here or just reduced the number of streams by half. It's possible the same failure happens if I try again with 100 tables instead of 50.
- On restart, materialize is not really responsive. select queries and even show table queries just hang indefinitely. So there's still stuff to shake out.
